crates/txpool/src/listener.rs

```diff
@@ -0,0 +1,89 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+use crate::error::Error;
+use std::{
+    fmt::{Debug, LowerHex},
+    sync::Arc,
+};
+
+/// Transaction pool listener.
+///
+/// Listener is being notified about status of every transaction in the pool.
+pub trait Listener<T> {
+    /// The transaction has been successfuly added to the pool.
+    /// If second argument is `Some` the transaction has took place of some other transaction
+    /// which was already in pool.
+    /// NOTE: You won't be notified about drop of `old` transaction separately.
+    fn added(&mut self, _tx: &Arc<T>, _old: Option<&Arc<T>>) {}
```

Comment:
Just for clarification, not a comment to change something: Why this needs to be a refcounted pointer instead of a plain old immutable reference?